### PR TITLE
Handle ITupleBinaryOperation in CFG

### PR DESCRIPTION
**Pull request by mavasani**

### CODE REVIEW

File: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs
Comment (jcouv) [inline src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs]: The ordering for `(a, b) == (c, d)` is `tempA = a; tempB = b; tempC = c; tempD = d; tempA == tempC && tempB == tempD`.
That last expression has a mini control-flow. This can skip some conversions and also some user-defined comparison operators. 
Do we expect to represent that in the CFG?

Comment (mavasani) [reply src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs]: Based on our discussion, filed https://github.com/dotnet/roslyn/issues/26969 as a follow-up issue.

Comment (jaredpar) [reply src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs]: Thanks for filing the issue. Definitely feels like something the CFG needs to express. There is no way we can expect consumers to be able to correctly infer the control flow + side effects here. Too complex. 

---
In reply to: [189291960](https://github.com/dotnet/roslyn/pull/26919#discussion_r189291960) [](ancestors = 189291960)

**Comment by mavasani** [comment]
@dotnet-bot retest windows_debug_spanish_unit32_prtest  please

File: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs]: >_evalStack.Push(Visit(operation.LeftOperand)); [](start = 12, length = 46)

If operands are tuple expressions, should we preserve those as we do for deconstruction, for example?

File: src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_ITupleBinaryOperatorExpression.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_ITupleBinaryOperatorExpression.cs]: > } [](start = 7, length = 2)

Please add tests when left and right are tuple operations and add some variations with control flow (given current implementation, no need to test more than one control flow construct on each side)

**Comment by AlekseyTs** [comment]
Done with review pass (iteration 1)

File: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs
Comment (AlekseyTs) [reply src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs]: If we do that, we probably should do that for the other side as well, but it looks like we are not doing this for deconstruction. Probably should do as well.

---
In reply to: [189724034](https://github.com/dotnet/roslyn/pull/26919#discussion_r189724034) [](ancestors = 189724034)

**Comment by mavasani** [comment]
@AlekseyTs addressed feedback

---
In reply to: [390796883](https://github.com/dotnet/roslyn/pull/26919#issuecomment-390796883) [](ancestors = 390796883)

**Review by AlekseyTs — APPROVED** [review]
LGTM (iteration 2)

**Comment by mavasani** [comment]
Ping @jcouv @333fred @dotnet/roslyn-compiler for second review

**Comment by 333fred** [comment]
@mavasani test failures look legitimate

**Review by 333fred — APPROVED** [review]
LGTM (iteration 2) once compile error is resolved.

**Comment by mavasani** [comment]
@333fred Thanks. Seems like merge commit introduced duplicate switch labels. Should be fixed now.

**Pull request merged**
